src/passes/OptimizeInstructions.cpp

```diff
@@ -2833,12 +2832,78 @@ struct OptimizeInstructions
             return c;
           }
         };
-        ifTrue = updateArm(ifTrue);
-        ifFalse = updateArm(ifFalse);
+        curr->ifTrue = updateArm(curr->ifTrue);
+        curr->ifFalse = updateArm(curr->ifFalse);
         un->value = curr;
         return replaceCurrent(un);
       }
     }
+
+    {
+      // Identical code on both arms can be folded out, e.g.
+      //
+      //  (select
+      //    (i32.eqz (X))
+      //    (i32.eqz (Y))
+      //    (Z)
+      //  )
+      // =>
+      //  (i32.eqz
+      //    (select
+      //      (X)
+      //      (Y)
+      //      (Z)
+      //    )
+      //  )
+      //
+      // Continue doing this while we can, noting the chain of moved expressions
+      // as we go, then do a single replaceCurrent() at the end.
+      SmallVector<Expression*, 1> chain;
```

Comment:
I did consider that, but it seemed less simple to me. I suppose it would be more efficient in memory, but it's extremely rare that there is more than 1 item here I suspect...